File: server/domain/entity/message.go
package entity

import (
	"time"
)

type Message struct {
	ID               string
	ChannelID        uint64
	Channel          Channel
	UserID           uint64
	User             User
	ReplyToMessageID string
	ReplyToMessage   *Message
	Content          string
	CreatedAt        time.Time
	UpdatedAt        time.Time
	DeletedAt        *time.Time
}

func NewMessageEntity(id string, channelID uint64, channel Channel, userID uint64, user User, replyToMessageID string, replyToMessage *Message, content string, createdAt, updatedAt time.Time, deletedAt *time.Time) *Message {
	return &Message{
		ID:               id,
		ChannelID:        channelID,
		Channel:          channel,
		UserID:           userID,
		User:             user,
		ReplyToMessageID: replyToMessageID,
		ReplyToMessage:   replyToMessage,
		Content:          content,
		CreatedAt:        createdAt,
		UpdatedAt:        updatedAt,
		DeletedAt:        deletedAt,
	}
}

File: client/src/repositories/message.ts
export interface CreateMessageParam {
  content: string;
}

interface User {
  id: number;
  name: string;
  image_url: string;
}

interface ReplyToMessage {
  id: string;
  user: User;
  content: string;
}

export interface Message {
  id: string;
  user: User;
  content: string;
  reply_to_message: ReplyToMessage | null;
  created_at: string;
  updated_at: string;
}

interface CreateMessageResponse {
  messages: Message[];
}

export interface IMessageRepository {
  createMessage: (
    channelId: number,
    param: CreateMessageParam
  ) => Promise<void>;
  getMessages: (channelId: number) => Promise<CreateMessageResponse>;
  getMessages$$key: (channelId: number) => string[];
}

export class MessageRepositoryImpl implements IMessageRepository {
  async createMessage(channelId: number, param: CreateMessageParam) {
    const res = await fetch(`/api/channel/${channelId}/message`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(param),
    });

    if (!res.ok) {
      throw new Error("Failed to create message");
    }
  }

  async getMessages(channelId: number) {
    const res = await fetch(`/api/channel/${channelId}/message`);

    return res.json();
  }

  getMessages$$key(channelId: number) {
    return ["getMessages", `channelId:${channelId}`];
  }
}
